[clang-tidy] Strict mode causes misleading redundant inline specifier warning

#73069 introduced a new warning [readability-redundant-inline-specifier](https://clang.llvm.org/extra/clang-tidy/checks/readability/redundant-inline-specifier.html). But when I try to fix the warnings in our project, when `Strict` is enabled, I'm getting compile issues by MSVC.

Following code is [compiling under MSVC](https://godbolt.org/z/6ajcMGWdP):

```cpp
namespace
{

class A
{
public:
  static inline const float test = 3.0F;
};

}

int main()
{
    A a;
    return 0;
};
```

But when I run clang-tidy over this code, via:
```sh
clang-tidy-18 -checks='readability-redundant-inline-specifier' -config='{CheckOptions: [{key: readability-redundant-inline-specifier.StrictMode, value: true}]}' test.cpp -- -std=c++17
```

I'm getting:
```
test.cpp:7:10: warning: variable 'test' has inline specifier but is implicitly inlined [readability-redundant-inline-specifier]
    7 |   static inline const float test = 3.0F;
      |          ^~~~~~
```

When I remove the `inline` Clang-Tidy is happy (also when I remove the anonymous namespace), but than MSVC complains:

```
example.cpp
<source>(7): error C2864: '`anonymous-namespace'::A::test': a static data member with an in-class initializer must have non-volatile const integral type or be specified as 'inline'
<source>(7): note: type is 'const float'
Compiler returned: 2
```

The documentation of the warning [C2864](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-errors-2/compiler-error-c2864) explicitly show an example with a floating point type - I have no idea why `inline` is required for floating point types and not for integral types, but as they have explicitly mentioned it, there must be a reason.

--- a/clang-tools-extra/clang-tidy/readability/RedundantInlineSpecifierCheck.cpp
+++ b/clang-tools-extra/clang-tidy/readability/RedundantInlineSpecifierCheck.cpp
@@ -47,9 +47,6 @@ AST_POLYMORPHIC_MATCHER_P(isInternalLinkage,
     return VD->isInAnonymousNamespace();
   llvm_unreachable("Not a valid polymorphic type");
 }
-
-AST_MATCHER(clang::VarDecl, hasInitialization) { return Node.hasInit(); }
-
 } // namespace
 
 static SourceLocation getInlineTokenLocation(SourceRange RangeLocation,
@@ -92,13 +89,15 @@ void RedundantInlineSpecifierCheck::registerMatchers(MatchFinder *Finder) {
 
   if (getLangOpts().CPlusPlus17) {
     const auto IsPartOfRecordDecl = hasAncestor(recordDecl());
-    Finder->addMatcher(varDecl(isInlineSpecified(),
-                               anyOf(allOf(isInternalLinkage(StrictMode),
-                                           unless(allOf(hasInitialization(),
-                                                        IsPartOfRecordDecl))),
-                                     allOf(isConstexpr(), IsPartOfRecordDecl)))
-                           .bind("var_decl"),
-                       this);
+    Finder->addMatcher(
+        varDecl(
+            isInlineSpecified(),
+            anyOf(allOf(isInternalLinkage(StrictMode),
+                        unless(allOf(hasInitializer(expr()), IsPartOfRecordDecl,
+                                     isStaticStorageClass()))),
+                  allOf(isConstexpr(), IsPartOfRecordDecl)))
+            .bind("var_decl"),
+        this);
   }
 }
 

--- a/clang-tools-extra/clang-tidy/readability/RedundantInlineSpecifierCheck.cpp
+++ b/clang-tools-extra/clang-tidy/readability/RedundantInlineSpecifierCheck.cpp
@@ -47,6 +47,9 @@ AST_POLYMORPHIC_MATCHER_P(isInternalLinkage,
     return VD->isInAnonymousNamespace();
   llvm_unreachable("Not a valid polymorphic type");
 }
+
+AST_MATCHER(clang::VarDecl, hasInitialization) { return Node.hasInit(); }
+
 } // namespace
 
 static SourceLocation getInlineTokenLocation(SourceRange RangeLocation,
@@ -88,12 +91,14 @@ void RedundantInlineSpecifierCheck::registerMatchers(MatchFinder *Finder) {
         this);
 
   if (getLangOpts().CPlusPlus17) {
-    Finder->addMatcher(
-        varDecl(isInlineSpecified(),
-                anyOf(isInternalLinkage(StrictMode),
-                      allOf(isConstexpr(), hasAncestor(recordDecl()))))
-            .bind("var_decl"),
-        this);
+    const auto IsPartOfRecordDecl = hasAncestor(recordDecl());
+    Finder->addMatcher(varDecl(isInlineSpecified(),
+                               anyOf(allOf(isInternalLinkage(StrictMode),
+                                           unless(allOf(hasInitialization(),
+                                                        IsPartOfRecordDecl))),
+                                     allOf(isConstexpr(), IsPartOfRecordDecl)))
+                           .bind("var_decl"),
+                       this);
   }
 }
 

--- a/clang-tools-extra/docs/ReleaseNotes.rst
+++ b/clang-tools-extra/docs/ReleaseNotes.rst
@@ -160,6 +160,10 @@ Changes in existing checks
   `AllowStringArrays` option, enabling the exclusion of array types with deduced
   length initialized from string literals.
 
+- Improved :doc:`readability-redundant-inline-specifier
+  <clang-tidy/checks/readability/redundant-inline-specifier>` check to properly
+  emit warnings for static data member with an in-class initializer.
+
 Removed checks
 ^^^^^^^^^^^^^^
 

--- a/clang-tools-extra/test/clang-tidy/checkers/readability/redundant-inline-specifier.cpp
+++ b/clang-tools-extra/test/clang-tidy/checkers/readability/redundant-inline-specifier.cpp
@@ -135,3 +135,17 @@ INLINE_MACRO()
 
 #define INLINE_KW inline
 INLINE_KW void fn10() { }
+
+namespace {
+class A
+{
+public:
+  static inline float test = 3.0F;
+  static inline double test2 = 3.0;
+  static inline int test3 = 3;
+
+  static inline float test4;
+  // CHECK-MESSAGES-STRICT: :[[@LINE-1]]:10: warning: variable 'test4' has inline specifier but is implicitly inlined [readability-redundant-inline-specifier]
+  // CHECK-FIXES-STRICT: static float test4;
+};
+}

